### PR TITLE
permissions-center: add a dot to status badge if some provider syncs failed.

### PR DESCRIPTION
**Pull request by sashaostrikov**
Test plan:
Local sg run and test.

### Demo

https://user-images.githubusercontent.com/94846361/227566230-db15292c-9058-4399-8e42-eaa08160a43e.mp4

Closes https://github.com/sourcegraph/sourcegraph/issues/49909

## App preview:

- [Web](https://sg-web-ao-ui-pc-add-warning-dot.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by kopancek** [comment]
I think I'd rather see `PARTIAL_SUCCESS` or similar status instead of an orange badge that hides the information behind a hover.

**Review by erzhtor — APPROVED** [review]

File: client/web/src/site-admin/permissions-center/PermissionsSyncJobNode.tsx [hunk 1]
```diff
@@ -102,24 +102,31 @@ export const JOB_STATE_METADATA_MAPPING: Record<PermissionsSyncJobState, JobStat
...
-        variant={JOB_STATE_METADATA_MAPPING[state].badgeVariant}
-    >
-        {state}
-    </Badge>
-)
+export const PermissionsSyncJobStatusBadge: React.FunctionComponent<PermissionsSyncJobStatusBadgeProps> = ({ job }) => {
+    const { state, cancellationReason, failureMessage, codeHostStates } = job
+    const needsWarning =
+        state === PermissionsSyncJobState.COMPLETED &&
+        codeHostStates.find(codeHostState => codeHostState.status) !== undefined
```
Comment (thenamankumar) [inline client/web/src/site-admin/permissions-center/PermissionsSyncJobNode.tsx:112]: useMemo

File: client/web/src/site-admin/permissions-center/PermissionsSyncJobNode.tsx [hunk 2]
```diff
@@ -102,24 +102,31 @@ export const JOB_STATE_METADATA_MAPPING: Record<PermissionsSyncJobState, JobStat
...
+    const warningMessage = needsWarning ? 'Some of provider syncs were not successful' : undefined
```
Comment (thenamankumar) [inline client/web/src/site-admin/permissions-center/PermissionsSyncJobNode.tsx:113]: `1/3 provider syncs were not successful`

File: client/web/src/site-admin/permissions-center/PermissionsSyncJobNode.tsx (hunk 1 above)
Comment (thenamankumar) [reply client/web/src/site-admin/permissions-center/PermissionsSyncJobNode.tsx:112]:
```javascript
const failingCodeHostSyncs = useMemo(() => codeHostStates.filter(({status}) => status === CodeHostStatus.FAILED).length, [codeHostStates])

const needsWarning = state === PermissionsSyncJobState.COMPLETED && failingCodeHostSyncs > 0
```

File: client/web/src/site-admin/permissions-center/PermissionsSyncJobNode.tsx (hunk 2 above)
Comment (thenamankumar) [reply client/web/src/site-admin/permissions-center/PermissionsSyncJobNode.tsx:113]:
```
`${failingCodeHostSyncs}/${codeHostStates.length} provider syncs were not successful.`
```

**Review by thenamankumar — CHANGES_REQUESTED** [review]
.

**Comment by thenamankumar** [comment]
@sashaostrikov based on what @kopancek suggested, I think making the Completed badge fully orange rather than green is a better idea in case we have partial success. 

We can still show the message on hover but remove the orange dot.

**Review by thenamankumar — APPROVED** [review]

**Comment by thenamankumar** [comment]
Is the font color in dark mode for all the warning badges blackish and not white? 

image

**Comment by sashaostrikov** [comment]
> Is the font color in dark mode for the warning badges blackish and not white everywhere?

Yep, it's like that in both light/dark modes: https://storybook.sgdev.org/?path=/story/wildcard-badge--badges

**Comment by danielmarquespt** [comment]
The problem with just changing the color is that Color Blind users will have an hard time perceiving the difference

Screenshot 2023-03-28 at 16 07 24

Even though you can tell something is different about those two "completed" you can't tell which one is better than the other. 

I think we really need to change the label here.

@sashaostrikov

**Comment by mrnugget** [comment]
How about `PARTIAL` as a label?

**Comment by sashaostrikov** [comment]
@danielmarquespt @mrnugget 
I understand the concern. In order to introduce a new label (e.g. `PARTIAL`), I think we should also support filtering by this label (imagine user confusion when they see `PARTIAL` status, but cannot filter by it in a dropdown called `State` where any other state is listed except `PARTIAL`).

**Which means we should introduce some changes to our `worker` framework**, I can totally do it, and I think there is a way of making it nicely:
- add an error type which indicates partial success
- add related check [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/workerutil/worker.go?L402-420)

In this case we won't break any other code using worker, I'll just update the documentation and the code which uses new status (effectively only our worker at the moment).

**Here is the question**
Since we are basically introducing a new worker status, we need to make it as universal as possible, which includes its name and reason. `PARTIAL`/`PARTIAL_SUCCESS` sounds good to me and makes sense, because we clearly have a lot of batch-style processing in workers and having a partial success is really a thing. Regarding the name, anyway it will be a database/backend code representation, which can be made more concise or user-friendly in the UI of needed, so I'm happy with this wording: `PARTIAL`/`PARTIAL_SUCCESS`.

WDYT? As soon as it sounds good to you -- I'll implement it.

**Comment by mrnugget** [comment]
Hmmm, I don't think we need or should change the `worker` framework. We could add another column instead, right? We do have `Post` handlers in workers, so we could check the result and then mark something as partial result. Would leave worker alone but give us additional results.

**Comment by sashaostrikov** [comment]
@mrnugget 
Another column would make filtering weird, don't you think? It will become not only `WHERE status = 'COMPLETED'`, but in case of `PARTIAL` it would check for another column (maybe boolean `is_partial_success` or so).

But if we don't want to change how `worker` _works_, then it should be another column.

I like another column idea less, but it would be safer to implement, than changing the `worker` code.

I guess then I'll add a boolean column to the table.

**Comment by mrnugget** [comment]
> It will become not only `WHERE status = 'COMPLETED'`, but in case of `PARTIAL` it would check for another column (maybe boolean `is_partial_success` or so).

Is that so bad?

I don't want to kill your argument by saying "let's do it my way", but the reason why I'm hesitant to add another status to `worker` is that this case here is pretty unique to our situation and _technically_ the job did _complete_, right? It finished, it did its work, the problem is that the work didn't yield the result that we want. I think the `worker_status` column should be about the _job_ status and not whether the _result_ of that job is what we want, especially if we don't retry.

Example: you have a job that that calculates average session duration for users, every day. some days some users don't log in. couldn't calculate a session duration for those users. the job still succeeds, right? even if the result is partial?

**Comment by sashaostrikov** [comment]
> I think the worker_status column should be about the job status and not whether the result of that job is what we want, especially if we don't retry.

That makes a lot of sense, totally agree!

> Is that so bad?

Not really, feels a bit like diluting the status between 2 columns, but nothing much. Won't affect performance either.

Thanks for discussion, really quick and productive! 👍🏻 

Now I go and implement [this](https://github.com/sourcegraph/sourcegraph/issues/50089)

**Pull request merged**